File: 695/max-area-of-island.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int ans = 0;
        for(int i=0; i<grid.size(); i++) {
            for(int j=0; j<grid[i].size(); j++) {
                if(grid[i][j]!=0) {
                    int temp = helper(grid, i, j);
                    if(temp>ans) ans = temp;
                }
            }
        }
        return ans;
    }
    int total = 0;
    int helper(vector<vector<int>>&grid, int i, int j) {
        if(i<0 || j<0 || i>=grid.size() || j>=grid[0].size()) return 0;

        if(grid[i][j]==0) return 0;
        else grid[i][j] = 0;

        int ans = 1;
        ans += helper(grid, i+1, j);
        ans += helper(grid, i-1, j);
        ans += helper(grid, i, j+1);
        ans += helper(grid, i, j-1);
        return ans;
    }
};

File: 695/max-area-of-island.py
# 這題用簡單的 DFS 便能把「能走到的」都走到了！
# 先用 stack + while 迴圈來做
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        M, N = len(grid), len(grid[0]) # 了解長寬
        visited = [[False]*N for _ in range(M)] #走過不再走
        stack = [] # 用來 visiting()用的DFS資料結構
        def visiting(i, j):
            if i<0 or j<0 or i>=M or j>=N:
                return # 超過範圍，就略過
            if not visited[i][j] and grid[i][j]==1:
                stack.append((i,j)) # 能走的話
                visited[i][j] = True # 就過過去

        ans = 0
        for i in range(M): # 左手i
            for j in range(N): # 右手j
                if grid[i][j]==1 and not visited[i][j]:
                    visiting(i,j) # 能參訪，就參訪
                    now = 0
                    while len(stack)>0: # 繼續往4週參訪
                        i2,j2 = stack.pop()
                        now += 1
                        visiting(i2+1,j2)
                        visiting(i2-1,j2)
                        visiting(i2,j2+1)
                        visiting(i2,j2-1)
                    ans = max(ans, now)
        return ans
